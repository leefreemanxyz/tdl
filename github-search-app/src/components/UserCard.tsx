import {
  Card,
  Link,
  Typography,
  CardContent,
  CardActionArea,
  Avatar,
  CircularProgress,
  Grid,
  makeStyles,
  Chip,
} from "@material-ui/core";
import { useQuery } from "react-query";
import axios from "axios";
import { Alert } from "@material-ui/lab";
import { GitHub, Twitter, Web } from "@material-ui/icons";
import { API_URL } from "../const";

const getValidUrl = (url = "") => {
  let newUrl = window.decodeURIComponent(url);
  newUrl = newUrl.trim().replace(/\s/g, "");

  if (/^(:\/\/)/.test(newUrl)) {
    return `http${newUrl}`;
  }
  if (!/^(f|ht)tps?:\/\//i.test(newUrl)) {
    return `http://${newUrl}`;
  }

  return newUrl;
};

const useStyles = makeStyles((theme) => ({
  small: {
    width: theme.spacing(3),
    height: theme.spacing(3),
  },
  large: {
    width: theme.spacing(10),
    height: theme.spacing(10),
  },
  root: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    "& > *": {
      margin: theme.spacing(1),
    },
  },
}));

const getUserDetail = ({ queryKey }: { queryKey: any }) => {
  const [, { username }] = queryKey;
  return axios.get(`${API_URL}/user/${username}`);
};

function UserCard({ username }: { username: string }) {
  const classes = useStyles();
  const { data, isLoading, isError } = useQuery(
    ["userDetail", { username }],
    getUserDetail,
    {
      staleTime: 60 * 60 * 1000,
    }
  );
  const user = data?.data ?? false;

  if (isError) {
    return (
      <Card>
        <CardContent>
          <CardActionArea>
            <Alert severity="error">Something went wrong!</Alert>
          </CardActionArea>
        </CardContent>
      </Card>
    );
  }

  if (isLoading) {
    return (
      <Card>
        <CardContent>
          <CardActionArea>
            <CircularProgress />
          </CardActionArea>
        </CardContent>
      </Card>
    );
  }

  return (
    user && (
      <Card data-testid="user-result" className={classes.root}>
        <CardContent>
          <Grid container>
            <Grid
              container
              item
              xs={4}
              direction="column"
              alignItems="center"
              justify="center"
            >
              <Grid>
                <Avatar
                  className={classes.large}
                  alt=""
                  src={user.avatar_url}
                />
              </Grid>
              <Grid>
                <Typography data-testid={`username-${user.login}`}>
                  {user.login}
                </Typography>
              </Grid>
              <Grid>
                <Typography>{user.name}</Typography>
              </Grid>
              <Grid>
                <Typography>{user.location}</Typography>
              </Grid>
            </Grid>
            <Grid
              container
              item
              xs={8}
              direction="column"
              justify="space-between"
            >
              <Grid>
                <Typography variant="body1">
                  {user.bio || "No user bio ðŸ˜­"}
                </Typography>
              </Grid>
              <Grid>
                {user.company && <Chip label={`Works for: ${user.company}`} />}
                <Chip label={`Followers: ${user.followers}`} />
                <Chip label={`Following: ${user.following}`} />
                <Chip label={`Public repos: ${user.public_repos}`} />
                <Chip label={`Public gists: ${user.public_gists}`} />
              </Grid>
              <Grid>
                {user.twitter_username && (
                  <Link
                    data-testid={`twitter-${user.login}`}
                    href={`https://twitter.com/${user.twitter_username}`}
                  >
                    <Chip
                      clickable
                      icon={<Twitter />}
                      label={`@${user.twitter_username}`}
                    />
                  </Link>
                )}
                {user.blog && (
                  <Link
                    data-testid={`personal-${user.login}`}
                    href={getValidUrl(user.blog)}
                  >
                    <Chip clickable icon={<Web />} label="Personal site" />
                  </Link>
                )}
                {user.html_url && (
                  <Link
                    data-testid={`github-${user.login}`}
                    href={user.html_url}
                  >
                    <Chip clickable icon={<GitHub />} label="GitHub" />
                  </Link>
                )}
              </Grid>
            </Grid>
          </Grid>
        </CardContent>
      </Card>
    )
  );
}

export default UserCard;
