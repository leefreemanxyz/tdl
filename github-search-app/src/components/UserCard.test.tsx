import { act, render } from "@testing-library/react";
import UserCard from "./UserCard";
import axios from "axios";
import { QueryClient, QueryClientProvider } from "react-query";
import { ReactNode } from "react";
jest.mock("axios");

const queryClient = new QueryClient();

const Wrapper = ({ children }: { children: ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe("UserCard", () => {
  it("snapshot", async () => {
    axios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          avatar_url: "src/img.jpg",
          login: "leefreemanxyz",
          name: "Lee Freeman",
          location: "Amsterdam",
          bio: undefined,
          twitter_username: "leefreemanxyz",
          followers: 27,
          following: 28,
          public_repos: 29,
          public_gists: 30,
          html_url: "https://github.com/leefreemanxyz",
          blog: "https://leefreeman.xyz",
        },
      })
    );
    let result;
    await act(async () => {
      result = await render(
        <Wrapper>
          <UserCard username="leefreemanxyz" />
        </Wrapper>
      );
    });
    const { container } = result;
    expect(container).toMatchSnapshot();
  });
});
