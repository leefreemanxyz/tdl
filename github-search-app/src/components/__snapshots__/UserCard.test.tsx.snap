// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="user-result"
  >
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-4"
        >
          <div
            class="MuiGrid-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle makeStyles-large-2"
            >
              <img
                alt=""
                class="MuiAvatar-img"
                src="src/img.jpg"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
              data-testid="username-leefreemanxyz"
            >
              leefreemanxyz
            </p>
          </div>
          <div
            class="MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Lee Freeman
            </p>
          </div>
          <div
            class="MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Amsterdam
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-justify-xs-space-between MuiGrid-grid-xs-8"
        >
          <div
            class="MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              No user bio ðŸ˜­
            </p>
          </div>
          <div
            class="MuiGrid-root"
          >
            <div
              class="MuiChip-root"
            >
              <span
                class="MuiChip-label"
              >
                Followers: 27
              </span>
            </div>
            <div
              class="MuiChip-root"
            >
              <span
                class="MuiChip-label"
              >
                Following: 28
              </span>
            </div>
            <div
              class="MuiChip-root"
            >
              <span
                class="MuiChip-label"
              >
                Public repos: 29
              </span>
            </div>
            <div
              class="MuiChip-root"
            >
              <span
                class="MuiChip-label"
              >
                Public gists: 30
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              data-testid="twitter-leefreemanxyz"
              href="https://twitter.com/leefreemanxyz"
            >
              <div
                class="MuiButtonBase-root MuiChip-root MuiChip-clickable"
                role="button"
                tabindex="0"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiChip-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22.46 6c-.77.35-1.6.58-2.46.69.88-.53 1.56-1.37 1.88-2.38-.83.5-1.75.85-2.72 1.05C18.37 4.5 17.26 4 16 4c-2.35 0-4.27 1.92-4.27 4.29 0 .34.04.67.11.98C8.28 9.09 5.11 7.38 3 4.79c-.37.63-.58 1.37-.58 2.15 0 1.49.75 2.81 1.91 3.56-.71 0-1.37-.2-1.95-.5v.03c0 2.08 1.48 3.82 3.44 4.21a4.22 4.22 0 0 1-1.93.07 4.28 4.28 0 0 0 4 2.98 8.521 8.521 0 0 1-5.33 1.84c-.34 0-.68-.02-1.02-.06C3.44 20.29 5.7 21 8.12 21 16 21 20.33 14.46 20.33 8.79c0-.19 0-.37-.01-.56.84-.6 1.56-1.36 2.14-2.23z"
                  />
                </svg>
                <span
                  class="MuiChip-label"
                >
                  @leefreemanxyz
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              data-testid="personal-leefreemanxyz"
              href="https://leefreeman.xyz"
            >
              <div
                class="MuiButtonBase-root MuiChip-root MuiChip-clickable"
                role="button"
                tabindex="0"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiChip-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm-5 14H4v-4h11v4zm0-5H4V9h11v4zm5 5h-4V9h4v9z"
                  />
                </svg>
                <span
                  class="MuiChip-label"
                >
                  Personal site
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              data-testid="github-leefreemanxyz"
              href="https://github.com/leefreemanxyz"
            >
              <div
                class="MuiButtonBase-root MuiChip-root MuiChip-clickable"
                role="button"
                tabindex="0"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiChip-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
                  />
                </svg>
                <span
                  class="MuiChip-label"
                >
                  GitHub
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
